chain/network/src/peer/peer_actor.rs

```diff
@@ -151,37 +157,92 @@ impl PeerActor {
             peer_counter,
             routed_message_cache: LruCache::new(ROUTED_MESSAGE_CACHE_SIZE),
             throttle_controller,
+            protocol_buffers_supported: false,
+            force_encoding,
         }
     }
 
-    fn send_message(&mut self, msg: &PeerMessage) {
+    fn parse_message(&mut self, msg: &[u8]) -> anyhow::Result<PeerMessage> {
+        if let Some(e) = self.force_encoding {
+            return PeerMessage::deserialize(e, msg);
+        }
+        if self.peer_status == PeerStatus::Connecting {
+            if let Ok(msg) = PeerMessage::deserialize(crate::network_protocol::Encoding::Proto, msg)
+            {
+                self.protocol_buffers_supported = true;
+                return Ok(msg);
+            }
+            match PeerMessage::deserialize(crate::network_protocol::Encoding::Borsh, msg) {
+                Ok(msg) => Ok(msg),
+                Err(err) => {
+                    self.send_message_or_log(&PeerMessage::HandshakeFailure(
+                        self.my_node_info.clone(),
+                        HandshakeFailureReason::ProtocolVersionMismatch {
+                            version: PROTOCOL_VERSION,
+                            oldest_supported_version: PEER_MIN_ALLOWED_PROTOCOL_VERSION,
+                        },
+                    ));
+                    Err(err)
+                }
+            }
+        } else if self.protocol_buffers_supported {
+            PeerMessage::deserialize(crate::network_protocol::Encoding::Proto, msg)
+        } else {
+            PeerMessage::deserialize(crate::network_protocol::Encoding::Borsh, msg)
+        }
+    }
+
+    fn send_message_or_log(&mut self, msg: &PeerMessage) {
+        if let Err(err) = self.send_message(msg) {
+            debug!(target: "network", "send_message(): {}", err);
```

Comment:
Hm, I feel this should be `error` rather than debug? Serialization failure means we have a rather bad bug internally, right? That's why I am very doubtful about using anyhow: from the casual reading, we might see two kinds of errors here: either serialization failed, or we get some IO error when writing to a socket. The second one should be debug (peers might be malicious), but the first one clearly signals a bug in our code, and should not pass silently. Given that this is just `anyhow::Error`, its impossible to undrestand what *can* actually go wrong here, withoug reading the transitive closure of the called code.

Reply:
yeah, so:
- deserialization error should just be logged (definitely not error, more like warn/info)
- serialization error is an implementation bug (I'd go with a panic! from within send_message_with_encoding)
- error while writing to a descriptor should be a typed error

Reply:
SGTM! It's interesting that now, I think, we can log serialization error at the point where it occurs, rather than propagating it upwards.

Reply:
done